ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+  def register_workflow(Class, name: nil) -> void
+
+  # Register an activity with an optional :name
+  def register_activity(Class, name: nil) -> void
+
+  # Start the worker and block until the process exits
+  def run() -> void
+
+  # Start the worker asynchronously
+  def start() -> void
+
+  # Gracefully stop the worker (wait for all running workflow/activities to finish)
+  def stop() -> void
+end
+```
+
+Internally each worker will create a thread for task polling as well as a thread pool of a given
+size (passed to the initializer as an optional `:threads` argument). Then each activity task will be
+dispatched to an available thread (via an in-memory queue). To make sure a polling worker always has
+processing capacity a polling request will only be dispatched if there is an available thread in the
+pool (in other words the pool will have no buffer).
+
+A very similar model will be used for workflow tasks, with an added complexity of wrapping each
+workflow in its own Fiber (inside a thread). This serves two purposes:
+
+1. A workflow code can be paused and resumed during the history replay
+2. Sticky workflows can be left paused while waiting for the next workflow task
+
+Given that Fibers are very lightweight this will allow us to keep many stick workflows efficiently
+in memory and manage accordingly to the worker configuration.
+
+Here's a diagram to demonstrate the whole setup:
+
+<img src="./ruby_threads.png" width="400" align="center" alt="Temporal" />
+
+### Alternatives considered
+
+We've decided against using the `Client` to initialise the `Worker` (something that Go and Python
+SDKs are doing). It seemed like a wrong abstraction and would require making many attributes public
+on the client to make sure the worker can access them (expanding the public API of a client). Using
+a configuration object solves this and opens up further optimisation opportunities (that will be
+proposed in the 2nd phase).
+
+We can potentially use the reactor pattern via Async gem. This has the benefit of giving extra
+performance, however it would drastically reduce the performance unless activities are written in a
+reactor-friendly way. This can be further added as an optional approach to apps already using Async,
+but doesn't feel like a great default option.
+
+Additionally the SDK users will want to leverage multiple CPUs by spawning multiple worker
+processes. We have agreed that this part should be left out of the Ruby SDK and handled by
+developers themselves. Potentially this functionality can be added as a separate Ruby gem.
+
+
+## Activities
+
+Each activity is going to be represented by its own class. This decision comes with a few benefits:
+
+- There's no need to pollute global scope with activity methods (see "Global Scope" section)
+- There's no need to inject global context for communicating with Temporal. It will be available
+  through a private variable on the class instance (`#activity`)
```

Comment:
I think thread local is a good fit in this model, it allows using activity APIs outside of the single class without needing to pass the context around, I'd encourage you to reconsider.

Reply:
Do you see this as a blocking concern? I don't see how this would change anything, but just adds a bit of sugar coating that some SDK users might want. This is fully compatible with the current proposal and can be added later. Wouldn't you agree?